core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java

```diff
@@ -390,4 +416,80 @@ private static boolean isSplitLocal(List<HostAddress> splitAddresses, HostAddres
         }
         return false;
     }
+
+    static class QueueSizeAdjuster
+    {
+        private static final int MAX_ADJUSTMENT = 32;
+        private static final int MIN_ADJUSTMENT = 1;
+        private final Map<String, QueueStats> queueStats;
+        private final Map<String, Integer> sizeAdjustment;
+        private final Set<String> adjustedNodes;
+
+        QueueSizeAdjuster()
+        {
+            queueStats = new HashMap<>();
+            adjustedNodes = new HashSet<>();
+            sizeAdjustment = new HashMap<>();
+        }
+
+        public void update(List<RemoteTask> existingTasks)
+        {
+            adjustedNodes.clear();
+            for (RemoteTask task : existingTasks) {
+                TaskStats taskStats = task.getTaskInfo().getStats();
+                QueueStats newQueueStats = new QueueStats(
+                        taskStats.getCompletedDrivers(),
+                        0,
+                        taskStats.getRunningDrivers(),
+                        taskStats.getQueuedDrivers());
+                queueStats.compute(task.getNodeId(), (key, oldQueueStats) -> {
+                    if (oldQueueStats == null) {
+                        return newQueueStats;
+                    }
+                    else {
+                        return oldQueueStats.merge(newQueueStats);
+                    }
+                });
+            }
+        }
+
+        public int getQueueSizeAdjustment(String nodeId)
+        {
+            int currentAdjustment = sizeAdjustment.getOrDefault(nodeId, 1);
+            QueueStats stats = queueStats.get(nodeId);
+
+            if (adjustedNodes.contains(nodeId) || stats == null) {
+                return currentAdjustment;
+            }
+            adjustedNodes.add(nodeId);
+
+            if (currentAdjustment < MAX_ADJUSTMENT && (stats.runningSplits() > 0 || stats.completedSplits() > 0)
+                    && (stats.queuedSplits() == 0
+                    || stats.completedSplitsDiff() >= stats.queuedSplits()
+                    || stats.queuedSplits() <= stats.runningSplits())) {
+                currentAdjustment *= 2;
+                sizeAdjustment.put(nodeId, currentAdjustment);
+            }
+            else if (currentAdjustment > MIN_ADJUSTMENT &&
+                    ((stats.completedSplitsDiff() * 2) < stats.queuedSplits()
+                            && stats.queuedSplits() > (stats.runningSplits() * 2))) {
+                currentAdjustment /= 2;
+                sizeAdjustment.put(nodeId, currentAdjustment);
```

Comment:
you can hoist `sizeAdjustment.put(nodeId, currentAdjustment); `outside of the `if else`

Reply:
Not always the adjustment is changed. I wanted to update it only if there is a difference.